File: src/main/org/nlogo/prim/etc/_shapes.java
package org.nlogo.prim.etc;

import java.util.List;

import org.nlogo.api.LogoListBuilder;
import org.nlogo.api.Shape;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.Reporter;
import org.nlogo.nvm.Syntax;

public final strictfp class _shapes
    extends Reporter {
  @Override
  public Syntax syntax() {
    return Syntax.reporterSyntax(Syntax.TYPE_LIST);
  }

  @Override
  public Object report(Context context) {
    List<Shape> shapes = world.turtleShapeList().getShapes();
    LogoListBuilder result = new LogoListBuilder();
    for (Shape shape : shapes) {
      result.add(shape.getName());
    }
    return result.toLogoList();
  }
}

File: src/main/prim/etc/_shapes.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.etc

import org.nlogo.core.{ Syntax }
import org.nlogo.api.{ LogoList, LogoListBuilder }
import org.nlogo.nvm.{ Context, Reporter }

class _shapes extends Reporter {
  override def syntax =
    Syntax.reporterSyntax(
      ret = Syntax.ListType)
  override def report(context: Context): LogoList = {
    val result = new LogoListBuilder
    import collection.JavaConverters._
    for(shape <- world.turtleShapeList.getShapes.asScala)
      result.add(shape.getName)
    result.toLogoList
  }
}
